Custom Blockchains > Add Existing Pallet (#91)
polkadot-developers/polkadot-docs: develop/blockchains/custom-blockchains/add-existing-pallets.md

- The locations and versions of the pallets that are to be imported as dependencies for the runtime
- The features in each pallet that should be enabled when compiling the native Rust binary. By enabling the standard (`std`) feature set from each pallet, you ensure that the runtime includes the functions, types, and primitives necessary for the native build, which are otherwise excluded when compiling the Wasm binary

!!! note
    For information about adding dependencies in `Cargo.toml` files, see the [Dependencies](https://doc.rust-lang.org/cargo/guide/dependencies.html){target=\_blank} page in the Cargo documentation. For information about enabling and managing features from dependent packages, see the [Features](https://doc.rust-lang.org/cargo/reference/features.html){target=\_blank} section in the Cargo documentation.

## Dependencies for a New Pallet

> To add the dependencies for a new pallet to the runtime, you need to modify the `Cargo.toml` file by adding a new line into the `[workspace.dependencies]` section with the pallet you want to add. This pallet definition might look like:

Reviewer: ```suggestion
To add the dependencies for a new pallet to the runtime, you must modify the `Cargo.toml` file by adding a new line into the `[workspace.dependencies]` section with the pallet you want to add. This pallet definition might look like:
```

```toml
pallet-example = { version = "4.0.0-dev", default-features = false }
```

This line imports the pallet-example crate as a dependency and specifies the following:

- **`version`** - the specific version of the crate to import
- **`default-features`** - determines the behavior for including pallet features when compiling the runtime with standard Rust libraries

!!! note

>     If you’re importing a pallet that isn’t available on [`crates.io`](https://crates.io/){target=\_blank}, you can specify the location of the pallet (either locally or from a remote repository) by using the `git` or `path` key. For example:

Reviewer: ```suggestion
    If you’re importing a pallet that isn’t available on [`crates.io`](https://crates.io/){target=\_blank}, you can specify the pallet's location (either locally or from a remote repository) by using the `git` or `path` key. For example:
```

    ```toml
    pallet-example = { 
        version = "4.0.0-dev",
        default-features = false,
        git = "INSERT_PALLET_REMOTE_URL",
    }
    ```

    In this case, replace `INSERT_PALLET_REMOTE_URL` with the correct repository URL. For local paths, use the path key like so:

    ```toml
    pallet-example = { 
        version = "4.0.0-dev",
        default-features = false,
        path = "INSERT_PALLET_RELATIVE_PATH",
    }
    ```

>     Ensure to substitute `INSERT_PALLET_RELATIVE_PATH` with the appropriate local path to the pallet.

Reviewer: ```suggestion
    Ensure that you substitute `INSERT_PALLET_RELATIVE_PATH` with the appropriate local path to the pallet.
```

To ensure that the new dependencies resolve correctly for the runtime, you can run the following command:

```bash
cargo check --release
```

## Config Trait for Pallets

Every Polkadot SDK pallet defines a Rust trait called `Config`. This trait specifies the types and parameters that the pallet needs to integrate with the runtime and perform its functions. The primary purpose of this trait is to act as an interface between this pallet and the runtime in which it is embedded. A type, function, or constant in this trait is essentially left to be configured by the runtime that includes this pallet.

Consequently, a runtime that wants to include this pallet must implement this trait.

You can inspect any pallet’s `Config` trait by reviewing its Rust documentation or source code. The `Config` trait ensures the pallet has access to the necessary types (like events, calls, or origins) and integrates smoothly with the rest of the runtime.

At its core, the `Config` trait typically looks like this:

```rust
--8<-- 'code/develop/blockchains/custom-blockchains/add-existing-pallets/pallet-basic-config-trait.rs'
```

> This basic structure shows that every pallet must define certain types, such as `RuntimeEvent` and `WeightInfo`, to be functional within the runtime. The actual implementation can vary depending on the pallet’s specific needs.

Reviewer: ```suggestion
This basic structure shows that every pallet must define certain types, such as `RuntimeEvent` and `WeightInfo`, to function within the runtime. The actual implementation can vary depending on the pallet’s specific needs.
```

??? "Example - Utility Pallet"

>       For instance, in the [`utility pallet`](https://github.com/paritytech/polkadot-sdk/tree/master/substrate/frame/utility){target=\_blank}, the Config trait is implemented with the following types:

Reviewer: ```suggestion
      For instance, in the [`utility pallet`](https://github.com/paritytech/polkadot-sdk/tree/master/substrate/frame/utility){target=\_blank}, the `Config` trait is implemented with the following types:
```

      ```rust
      --8<-- 'code/develop/blockchains/custom-blockchains/add-existing-pallets/utility-pallet-config-trait.rs'
      ```

     This example shows how the `Config` trait defines types like `RuntimeEvent`, `RuntimeCall`, `PalletsOrigin`, and `WeightInfo`, which the pallet will use when interacting with the runtime.

> ## Parameters Configuration for Pallets

Reviewer: ```suggestion
## Parameter Configuration for Pallets
```

Traits in Rust define shared behavior, and within the Polkadot SDK, they allow runtimes to integrate and utilize a pallet's functionality by implementing its associated configuration trait and parameters. Some of these parameters may require constant values, which can be defined using the [`parameter_types!`](https://paritytech.github.io/polkadot-sdk/master/frame_support/macro.parameter_types.html){target=\_blank} macro. This macro simplifies development by expanding the constants into the appropriate struct types with functions that the runtime can use to access their types and values in a consistent manner.

For example, the following code snippet shows how the solochain template configures certain parameters through the [`parameter_types!`]({{ dependencies.polkadot_sdk_solochain_template.repository_url }}/blob/v0.0.2/runtime/src/lib.rs#L138){target=\_blank} macro in the `runtime/lib.rs` file:

```rust
--8<-- 'code/develop/blockchains/custom-blockchains/add-existing-pallets/parameter-types-example.rs'
```

## Pallet Config in the Runtime

> To integrate a new pallet into the runtime, you need to implement its `Config` trait in the `runtime/lib.rs` file. This is done by specifying the necessary types and parameters in Rust, as shown below:

Reviewer: ```suggestion
To integrate a new pallet into the runtime, you must implement its `Config` trait in the `runtime/lib.rs` file. This is done by specifying the necessary types and parameters in Rust, as shown below:
```
